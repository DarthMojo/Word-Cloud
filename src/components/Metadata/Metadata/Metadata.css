.Metadata {
    display: flex;
    flex-direction:row;
    align-items: center;
    justify-content: center;
    margin-top:20px;
}
.Metadata-Child{
    flex-grow: 1;
    justify-content: center;
}
