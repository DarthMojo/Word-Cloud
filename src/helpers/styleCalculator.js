/**
 * 1. Each topic should have one of 6 different text sizes,
 * with the most popular topics largest, and least popular smallest
 *
 * 2.A topic with a sentiment score > 60 should be displayed in green
 *
 * 3.A topic with a sentiment score < 40 should be displayed in red
 *
 * 4.Other topics should be displayed in grey
 */
import type Dimension from "../config/types";
import { sentimentScoreMin,
    sentimentScoreMax,
    defaultColor,
    size } from "../config/static";


/**
 *
 * @type {{getTopicSize: (function(*, *)),
 * getTopicColor: (function(*)),
 * windowSize: (function(*=)),
 * rotation: (function(): number)}}
 */
const styleCalculator = {
    /**
     *
     * @param wordPopularity
     * @param popularityRange
     * @return {number}
     */
    getTopicSize: (wordPopularity, popularityRange): number => {
        const unitRange = ((popularityRange.max - popularityRange.min) / size.max) || 0;
        const sizeValue: number = Array.from({ length: size.max }, (v, k): number => k + 1)
            .find((sizeIndex: number): number =>
                wordPopularity <= popularityRange.min + (sizeIndex * unitRange));
        return Math.round((sizeValue / size.max) * 100);
    },
    /**
     *
     * @param sentimentScore
     * @return String
     */
    getTopicColor: (sentimentScore): string => {
        if (sentimentScore > sentimentScoreMax.score) {
            return sentimentScoreMax.color;
        } else if (sentimentScore < sentimentScoreMin.score) {
            return sentimentScoreMin.color;
        }
        return defaultColor;
    },
    /**
     *
     * @param factor
     * @return {{width: number, height: number}}
     */
    windowSize: (factor = 1): Dimension => {
        return {
            width: Math.round(window.innerWidth * factor),
            height: Math.round(window.innerHeight * factor)
        };
    },
    /**
     *@return {number}
     */
    rotation: (): number => (~~(Math.random() * size.max) - size.max * 0.5) * 30
};
export default styleCalculator;
