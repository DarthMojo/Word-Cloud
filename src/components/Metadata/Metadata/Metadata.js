import React from "react";
import type { Element } from "react";
import getChartData from "../../../helpers/utils/getChartData";
import {
    pageTypeTitle,
    sentimentScoreTitle,
    popularityOverTimeTitle } from "../../../config/static";
import DataBarChart from "../Components/DataBarChart/DataBarChart";
import DataPieChart from "../Components/DataPieChart/DataPieChart";
import WordLabel from "../Components/WordLabel/WordLabel";
import BackButton from "../../BackButton/BackButton";

import "./Metadata.css";

type MetadataType = {
    wordData: Object,
    dimensions: Object,
    onBackButtonClick: Function
};

/**
 *
 * @param wordData
 * @param dimensions
 * @param onBackButtonClick
 * @return {Element}
 * @constructor
 */
const Metadata = ({ wordData, dimensions, onBackButtonClick }: MetadataType): Element<any> => {
    return (<div>
        <BackButton onButtonClick={onBackButtonClick} />
        <div className={"Metadata"}>
            <div className={"Metadata-Child"}>
                <WordLabel
                    wordColor={wordData.color}
                    wordLabel={wordData.label}
                    dimensions={dimensions}/>
                <DataPieChart
                    data={getChartData(wordData.sentiment)}
                    dimensions={dimensions}
                    chartTitle={sentimentScoreTitle}/>
            </div>
            <div className={"Metadata-Child"}>
                <DataBarChart
                    data={getChartData(wordData.pageType)}
                    dimensions={dimensions}
                    chartTitle={pageTypeTitle}/>
                <DataBarChart
                    data={wordData.days.map((day): Object => ({
                        x: new Date(day.date).toDateString(),
                        y: day.volume
                    }))}
                    dimensions={dimensions}
                    chartTitle={popularityOverTimeTitle}/>
            </div>
        </div>
    </div>
    );
};

export default Metadata;
