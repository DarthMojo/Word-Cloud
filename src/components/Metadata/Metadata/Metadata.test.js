/* eslint-disable no-undef */
import React from "react";
import Metadata from "./Metadata";
import data from "../../../config/data/topics.json";
import { shallowToJson } from "enzyme-to-json";
import styleCalculator from "../../../helpers/styleCalculator";

test("render Metadata()", () => {
    const dimensions = styleCalculator.windowSize();

    const onButtonClick = () => {};

    const wrapper = shallow(
        <Metadata
            dimensions={dimensions}
            wordData={data.topics.find((wordItem): Object<*> =>
                wordItem.id === "1751295897__Berlin")}
            onBackButtonClick={onButtonClick}
        />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
});
