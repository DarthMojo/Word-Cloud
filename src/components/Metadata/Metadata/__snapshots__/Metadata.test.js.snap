// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Metadata() 1`] = `
<div>
  <BackButton
    onButtonClick={[Function]}
  />
  <div
    className="Metadata"
  >
    <div
      className="Metadata-Child"
    >
      <WordLabel
        dimensions={
          Object {
            "height": 768,
            "width": 1024,
          }
        }
        wordLabel="Berlin"
      />
      <DataPieChart
        chartTitle="Sentiment Breakdown"
        data={
          Array [
            Object {
              "x": "negative",
              "y": 3,
            },
            Object {
              "x": "neutral",
              "y": 133,
            },
            Object {
              "x": "positive",
              "y": 29,
            },
          ]
        }
        dimensions={
          Object {
            "height": 768,
            "width": 1024,
          }
        }
      />
    </div>
    <div
      className="Metadata-Child"
    >
      <DataBarChart
        chartTitle="Page Hits"
        data={
          Array [
            Object {
              "x": "blog",
              "y": 17,
            },
            Object {
              "x": "facebook",
              "y": 56,
            },
            Object {
              "x": "forum",
              "y": 22,
            },
            Object {
              "x": "general",
              "y": 5,
            },
            Object {
              "x": "image",
              "y": 0,
            },
            Object {
              "x": "news",
              "y": 26,
            },
            Object {
              "x": "review",
              "y": 1,
            },
            Object {
              "x": "twitter",
              "y": 35,
            },
            Object {
              "x": "video",
              "y": 3,
            },
          ]
        }
        dimensions={
          Object {
            "height": 768,
            "width": 1024,
          }
        }
      />
      <DataBarChart
        chartTitle="Popularity Stats Over Time"
        data={
          Array [
            Object {
              "x": "Fri Jun 06 2014",
              "y": 22,
            },
            Object {
              "x": "Wed Jun 04 2014",
              "y": 43,
            },
            Object {
              "x": "Mon Jun 09 2014",
              "y": 0,
            },
            Object {
              "x": "Sat Jun 07 2014",
              "y": 12,
            },
            Object {
              "x": "Sun Jun 08 2014",
              "y": 11,
            },
            Object {
              "x": "Tue Jun 03 2014",
              "y": 39,
            },
            Object {
              "x": "Thu Jun 05 2014",
              "y": 38,
            },
          ]
        }
        dimensions={
          Object {
            "height": 768,
            "width": 1024,
          }
        }
      />
    </div>
  </div>
</div>
`;
